torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+    parser.parse(ttir)
+    return ops
+
+
+# Given a triton kernel and the arguments for this kernel, this function
+# retrieves the TTIR converted version of the kernel from Triton's API.
+# Parsing the TTIR, it creates a control flow graph, detects all sinks
+# from a predefined list of sinks. NOTE: What if triton exposed this MemWrite Trait
+# From each sink, it traverses the CFG backwards to identify all the input
+# pointers that are mutated
+def identify_mutated_tensors(kernel, kwargs):
+    from triton.compiler.code_generator import ast_to_ttir
     from triton.runtime.autotuner import Autotuner
+    from triton.runtime.jit import JITFunction
+
+    import torch
+    from torch._subclasses.fake_tensor import FakeTensor
 
     if isinstance(kernel, Autotuner):
+        if len(kernel.configs) > 0:
+            # If we are autotuning, then it doesn't matter which version gets
+            # picked for tracing purposes, so lets pick the first one
+            kwargs = {**kwargs, **kernel.configs[0].kwargs}
         kernel = kernel.fn
 
-    infos = {name: MutationInfo() for name in tensors}
-    tracker = MutationTracker(infos)
-    tracker.visit(kernel.parse())
+    assert isinstance(kernel, JITFunction)
+
+    # Drop the keys
+    # Replace all SymExprs with a regular value for TTIR generation
+    # Replace all FakeTensor with real tensors
+    # These replacements are needed for triton's type, key and config functions
+    args: List[Any] = []
+    for a in kwargs.values():
+        if isinstance(a, (torch.SymInt, torch.SymFloat, torch.SymBool)):
+            args.append(2)
+        elif isinstance(a, FakeTensor):
+            args.append(torch.empty(2, dtype=a.dtype))
+        else:
+            args.append(a)
+
+    specialization = kernel._get_config(*args)
+    constants = {i: arg for i, arg in enumerate(args) if not isinstance(arg, Tensor)}
+    debug = None
+    target = None
+
+    # Build kernel signature -- doesn't include constexpr arguments.
+    signature = {
+        i: kernel._type_of(kernel._key_of(arg))
+        for i, arg in enumerate(args)
+        if i not in kernel.constexprs
+    }
+
+    try:
+        ttir_module = ast_to_ttir(
+            kernel, signature, specialization, constants, debug, target
+        )
+        ops = parse_ttir(str(ttir_module), kwargs)
+        if ops is None:
+            raise Exception("Parsing TTIR failed")
+    except Exception as e:
+        import traceback
+
+        log.debug(
+            "Encountered an exception in identify_mutated_tensors, assuming every input is mutated"
+        )
+        log.debug(
+            "".join(
+                traceback.TracebackException.from_exception(e).format()  # noqa: G001
+            )
+        )
+        return [key for key, value in kwargs.items() if isinstance(value, Tensor)]
+
+    # Name of mutation op to mutated parameter indices
+    MUTATION_OPS = {"tt.store": [0], "tt.atomic_cas": [0], "tt.atomic_rmw": [0]}
```

Comment:
Is there any particular IR reference you're using to find the list of mutation ops?

Reply:
https://github.com/openai/triton/blob/7438dce18d296bb306f84e6a71ea6ee22e59535b/include/triton/Dialect/Triton/IR/TritonOps.td#L213

file contains MemWrite trait for all write operations.
Print, Assert, Store, AtomicRMWOp, AtomicCASOp are all the write operations. For our purposes we dont care about print and assert. I'll add a link.